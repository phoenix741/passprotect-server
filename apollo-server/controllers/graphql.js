import { GraphQLScalarType } from 'graphql'
import { Kind } from 'graphql/language'

const GraphQLScalarDate = new GraphQLScalarType({
  name: 'Date',
  description: 'Date in format Timestamp',
  serialize (value) {
    return value.getTime()
  },
  parseValue (value) {
    return new Date(value)
  },
  parseLiteral (ast) {
    if (ast.kind === Kind.INT) {
      return parseInt(ast.value, 10) // ast value is always in string format
    }
    return null
  }
})

export const resolvers = {
  Date: GraphQLScalarDate
}
