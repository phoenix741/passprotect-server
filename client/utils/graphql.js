/* global __GRAPHQL_SCHEMA__ */

import Vue from 'vue'
import VueApollo from 'vue-apollo'
import { ApolloClient, createNetworkInterface, IntrospectionFragmentMatcher } from 'apollo-client'
import { getAuthHeader } from '../components/user/UserService'

Vue.use(VueApollo)

const schema = __GRAPHQL_SCHEMA__

const networkInterface = createNetworkInterface({
  uri: '/api/graphql',
  transportBatching: true
})

networkInterface.use([{
  applyMiddleware (req, next) {
    if (!req.options.headers) {
      req.options.headers = {} // Create the header object if needed.
    }
    req.options.headers = getAuthHeader()
    next()
  }
}])

const fragmentMatcher = new IntrospectionFragmentMatcher({
  introspectionQueryResultData: schema.data
})

export const apolloClient = new ApolloClient({
  networkInterface,
  fragmentMatcher,
  connectToDevTools: true
})

export const apolloProvider = new VueApollo({defaultClient: apolloClient})
