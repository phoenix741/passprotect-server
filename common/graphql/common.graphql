scalar Date

# Encrypted content
type EncryptedContent {
  # Content encrypted with salt
  content: String!
  # Authentification tag
  authTag: String!
}

# Encrypted content
input EncryptedContentInput {
  # Content encrypted with salt
  content: String!
  # Authentification tag
  authTag: String!
}

type Error {
  fieldName: String,
  message: String
}

type Errors {
  errors: [Error]!
}

schema {
  query: RootQuery
  mutation: RootMutation
}
