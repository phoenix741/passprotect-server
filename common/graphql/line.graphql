# Wallet line content encrypted
type EncryptedWalletLine {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContent!
}

# Wallet line content encrypted
input EncryptedWalletLineInput {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContentInput!
}

# Wallet
type WalletLine {
  # Id of the line
  _id: ID!
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLine!,
  # Owner of the list
  user: User!,
  # Last updated date of the line
  updatedAt: Date,
  # Revision (modification count) of the line
  _rev: Int
}

# Wallet line used as input
input WalletLineInput {
  # Id of the line, if not filled application create a new line
  _id: ID
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLineInput!,
  # Revision (modification count) of the line, used to ensure modification of the last line
  _rev: Int
}

extend type RootQuery {
  # Request the list of lines from the connected users.
  lines: [WalletLine],

  # Request a wallet line
  line(id: ID!): WalletLine,

	# Request all groups available on lines
	groups: [String!]
}

union CreateUpdateLineResult = WalletLine | Errors

extend type RootMutation {
  createUpdateLine(input: WalletLineInput!): CreateUpdateLineResult
  removeLine(id: ID): Errors
}
