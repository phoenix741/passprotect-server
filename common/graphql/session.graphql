# Information needed to connect to the server and retreive object from database.
input ConnectionInformationInput {
  # Login of the user
  username: String!,
  # Password used for connection
  password: String!
}

# Result of the connection
type ConnectionInformation {
  # Token to be used in connection information
  token: String
}

extend type RootQuery {
  # Request the current session
  session: User
}

union CreateSessionResult = ConnectionInformation | Errors

extend type RootMutation {
  createSession(input: ConnectionInformationInput!): CreateSessionResult,
  clearSession: Errors
}
