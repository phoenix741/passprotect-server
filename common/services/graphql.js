'use strict';

const { flatten } = require('lodash');
const fs = require('fs');
const path = require('path');
const { makeExecutableSchema } = require('graphql-tools');
const { graphql, introspectionQuery } = require('graphql');

const typeDefs = flatten([
	fs.readFileSync(path.join(__dirname, '..', 'graphql', 'common.graphql'), 'utf-8'),
	fs.readFileSync(path.join(__dirname, '..', '..', 'common', 'graphql', 'user.graphql'), 'utf-8'),
	fs.readFileSync(path.join(__dirname, '..', '..', 'common', 'graphql', 'line.graphql'), 'utf-8'),
	fs.readFileSync(path.join(__dirname, '..', '..', 'common', 'graphql', 'session.graphql'), 'utf-8'),
	fs.readFileSync(path.join(__dirname, '..', '..', 'common', 'graphql', 'transaction.graphql'), 'utf-8')
]);

const schema = exports.schema = makeExecutableSchema({typeDefs});

exports.graphqlPromise = graphql(schema, introspectionQuery);
