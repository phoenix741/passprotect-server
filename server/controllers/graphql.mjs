import { flatten, merge } from 'lodash'
import fs from 'fs'
import path from 'path'
import debug from 'debug'
import { graphqlExpress } from 'graphql-server-express'
import { makeExecutableSchema } from 'graphql-tools'
import { GraphQLScalarType } from 'graphql'
import { Kind } from 'graphql/language'

import { typeDefs as userTypeDefs, resolvers as userResolvers } from './user'
import { typeDefs as lineTypeDefs, resolvers as lineResolvers } from './line'
import { typeDefs as sessionTypeDefs, resolvers as sessionResolvers } from './session'
import { typeDefs as transactionTypeDefs, resolvers as transactionResolvers } from './transaction'

const log = debug('App:Controllers:GraphQL')

log('Load common type definitions')
const typeDefs = flatten([
  fs.readFileSync(path.join(__dirname, '..', '..', 'common', 'graphql', 'common.graphql'), 'utf-8'),
  userTypeDefs,
  lineTypeDefs,
  sessionTypeDefs,
  transactionTypeDefs
])

const GraphQLScalarDate = new GraphQLScalarType({
  name: 'Date',
  description: 'Date in format Timestamp',
  serialize (value) {
    return value.getTime()
  },
  parseValue (value) {
    return new Date(value)
  },
  parseLiteral (ast) {
    if (ast.kind === Kind.INT) {
      return parseInt(ast.value, 10) // ast value is always in string format
    }
    return null
  }
})

const resolvers = merge({ Date: GraphQLScalarDate }, userResolvers, lineResolvers, sessionResolvers, transactionResolvers)

log('Create the graphql schema')
export const schema = makeExecutableSchema({ typeDefs, resolvers, logger: { log } })

const graphqlRouter = graphqlExpress((req, res) => ({
  schema,
  context: {
    user: req.user,
    res
  }
}))

export default graphqlRouter
