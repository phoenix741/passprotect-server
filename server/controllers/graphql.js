'use strict';

import { flatten, merge } from 'lodash';
import fs from 'fs';
import path from 'path';
import debug from 'debug';
import {authenticate} from 'server/utils/passport';
import { graphqlExpress } from 'graphql-server-express';
import { makeExecutableSchema } from 'graphql-tools';
import { GraphQLScalarType } from 'graphql';
import { Kind } from 'graphql/language';
import { SubscriptionManager } from 'graphql-subscriptions';

import {pubsub} from 'server/services/subscriptions';

import {typeDefs as userTypeDefs, resolvers as userResolvers} from './user';
import {typeDefs as lineTypeDefs, resolvers as lineResolvers} from './line';
import {typeDefs as sessionTypeDefs, resolvers as sessionResolvers} from './session';
import {typeDefs as transactionTypeDefs, resolvers as transactionResolvers, setupFunctions as transactionSetupFunctions} from './transaction';

const log = debug('App:Controllers:GraphQL');

log('Load common type definitions');
const typeDefs = flatten([
	fs.readFileSync(path.join(__dirname, '..', '..', 'common', 'graphql', 'common.graphql'), 'utf-8'),
	userTypeDefs,
	lineTypeDefs,
	sessionTypeDefs,
	transactionTypeDefs
]);

const GraphQLScalarDate = new GraphQLScalarType({
	name: 'Date',
	description: 'Date in format Timestamp',
	serialize(value) {
		return value.getTime();
	},
	parseValue(value) {
		return new Date(value);
	},
	parseLiteral(ast) {
		if (ast.kind === Kind.INT) {
			return parseInt(ast.value, 10); // ast value is always in string format
		}
		return null;
	}
});

const resolvers = merge({Date: GraphQLScalarDate}, userResolvers, lineResolvers, sessionResolvers, transactionResolvers);

const setupFunctions = merge(transactionSetupFunctions);

log('Create the graphql schema');
const schema = makeExecutableSchema({typeDefs, resolvers, logger: {log}});

export const subscriptionManager = new SubscriptionManager({schema, pubsub, setupFunctions});

const graphqlRouter = graphqlExpress((req, res) => ({
	schema,
	context: {
		user: req.user,
		res
	}
}));

export default [authenticate(), graphqlRouter];
