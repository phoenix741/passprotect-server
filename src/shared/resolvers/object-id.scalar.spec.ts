import { ObjectIDScalar } from './object-id.scalar';
import { ObjectID } from 'bson';
import { Kind } from 'graphql';

describe('ObjectIDScalar', () => {
  let scalar: ObjectIDScalar;

  beforeEach(() => {
    scalar = new ObjectIDScalar();
  });

  describe('serialize', () => {
    it('convert to hexstring', () => {
      expect(scalar.serialize(new ObjectID('5ce5b4a58382315c12788ea1'))).toMatchSnapshot();
    });
  });

  describe('parseValue', () => {
    it('parse value', () => {
      expect(scalar.parseValue('5ce5b4a58382315c12788ea1')).toMatchSnapshot();
      expect(scalar.parseValue(new ObjectID('5ce5b4a58382315c12788ea1'))).toMatchSnapshot();
    });
  });

  describe('parseLiteral', () => {
    it('parse literal', () => {
      expect(scalar.parseLiteral({
        kind: Kind.STRING,
        value: '5ce5b4a58382315c12788ea1',
      })).toMatchSnapshot();
      expect(scalar.parseLiteral({
        kind: Kind.FLOAT,
        value: 'float',
      })).toBeNull();
    });
  });
});
