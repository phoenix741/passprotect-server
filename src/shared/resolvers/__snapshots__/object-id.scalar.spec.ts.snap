// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectIDScalar parseLiteral parse literal 1`] = `"5ce5b4a58382315c12788ea1"`;

exports[`ObjectIDScalar parseValue parse value 1`] = `"5ce5b4a58382315c12788ea1"`;

exports[`ObjectIDScalar parseValue parse value 2`] = `"5ce5b4a58382315c12788ea1"`;

exports[`ObjectIDScalar serialize convert to hexstring 1`] = `"5ce5b4a58382315c12788ea1"`;
